File: src/info/meodinger/LabelPlusFX/Util/CZip.java
package info.meodinger.LabelPlusFX.Util;

import java.io.*;
import java.util.zip.*;


/**
 * Author: Meodinger
 * Date: 2021/5/31
 * Location: info.meodinger.LabelPlusFX.Util
 */
public class CZip {

    private final ZipOutputStream zip;

    public CZip(String zipFilePath) {
        ZipOutputStream zipOutputStream;
        try {
            BufferedOutputStream bufferedOutputStream = new BufferedOutputStream(new FileOutputStream(zipFilePath));
            zipOutputStream = new ZipOutputStream(bufferedOutputStream);
        } catch (IOException e) {
            CDialog.showException(e);
            throw new RuntimeException("Abort");
        }
        this.zip = zipOutputStream;
    }

    public void zip(File file, String path) throws IOException{
        if (file.isDirectory()) {
            path = path + "/";
            zip.putNextEntry(new ZipEntry(path));
            File[] files = file.listFiles();
            if (files != null) {
                for (File f : files) {
                    zip(f, path + f);
                }
            }
        } else {
            zip(new FileInputStream(file), path);
        }
    }

    public void zip(InputStream inputStream, String path) throws IOException {
        zip.putNextEntry(new ZipEntry(path));

        BufferedInputStream input = new BufferedInputStream(inputStream);

        byte[] chunk = new byte[1024 * 4];
        int len;
        while ((len = input.read(chunk)) != -1) {
            zip.write(chunk, 0, len);
        }
        input.close();
        zip.closeEntry();
    }

    public void close() throws IOException {
        zip.close();
    }
}

File: src/main/kotlin/info/meodinger/lpfx/util/CZip.kt
package info.meodinger.lpfx.util

import info.meodinger.lpfx.util.dialog.showException

import java.io.*
import java.util.zip.ZipEntry
import java.util.zip.ZipOutputStream


/**
 * Author: Meodinger
 * Date: 2021/7/29
 * Location: info.meodinger.lpfx.util
 */

/**
 * Zip files easier
 */
class CZip(zipFile: File) {
    private val zip: ZipOutputStream

    init {
        zip = try {
            ZipOutputStream(BufferedOutputStream(FileOutputStream(zipFile)))
        } catch (e: IOException) {
            showException(e)
            throw RuntimeException("Abort")
        }
    }

    @Throws(IOException::class)
    fun zip(file: File, path: String) {
        if (file.isDirectory) {
            val folder = "$path/"
            zip.putNextEntry(ZipEntry(folder))

            file.listFiles()?.forEach { f -> zip(f, folder + f) }
        } else {
            zip(FileInputStream(file), path)
        }
    }

    @Throws(IOException::class)
    fun zip(inputStream: InputStream, path: String) {
        zip.putNextEntry(ZipEntry(path))

        val input = BufferedInputStream(inputStream)
        val chunk = ByteArray(1024 * 4)
        var len: Int
        while (input.read(chunk).also { len = it } != -1) {
            zip.write(chunk, 0, len)
        }
        input.close()

        zip.closeEntry()
    }

    @Throws(IOException::class)
    fun zip(bytes: ByteArray, path: String) {
        zip.putNextEntry(ZipEntry(path))
        zip.write(bytes)
        zip.closeEntry()
    }

    @Throws(IOException::class)
    fun close() {
        zip.close()
    }
}